error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/PatternRules.java

```diff
@@ -0,0 +1,46 @@
+package tech.picnic.errorprone.refasterrules;
+
+import static com.google.common.base.Predicates.containsPattern;
+
+import com.google.common.base.Predicates;
+import com.google.errorprone.refaster.annotation.AfterTemplate;
+import com.google.errorprone.refaster.annotation.BeforeTemplate;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import tech.picnic.errorprone.refaster.annotation.OnlineDocumentation;
+
+/** Refaster rules related to code dealing with regular expressions. */
+@OnlineDocumentation
+final class PatternRules {
+  private PatternRules() {}
+
+  /** Prefer {@link Pattern#asPredicate()} over non-JDK alternatives. */
+  // XXX: This rule could also replace `s -> pattern.matcher(s).find()`, though the lambda
+  // expression may match functional interfaces other than `Predicate`. If we do add such a rule, we
+  // should also add a rule that replaces `s -> pattern.matcher(s).matches()` with
+  // `pattern.asMatchPredicate()`.
+  static final class PatternAsPredicate {
+    @BeforeTemplate
+    Predicate<CharSequence> before(Pattern pattern) {
+      return Predicates.contains(pattern);
+    }
+
+    @AfterTemplate
+    Predicate<String> after(Pattern pattern) {
+      return pattern.asPredicate();
```

Comment:
also here I'd personally argue that the Guava variant is more descriptive (nor do they recommend against using it)

I'm also not sure it's 100% equivalent as JDK uses `String` vs. Guava's `CharSequence`

Reply:
> also here I'd personally argue that the Guava variant is more descriptive (nor do they recommend against using it)

On descriptiveness I guess we'll just have to disagree :sweat_smile:. W.r.t. recommendations: while `contains` itself doesn't call it out, the returned [`com.google.common.base.Predicate`](https://guava.dev/releases/32.1.1-jre/api/docs/com/google/common/base/Predicate.html) is described as a legacy type that should be avoided whenever possible.


> I'm also not sure it's 100% equivalent as JDK uses `String` vs. Guava's `CharSequence`

Indeed it isn't; this is a Refaster limitation that unfortunately more of our rules suffer from. (Longer-term I plan to automatically flag and annotate such rules such that they can e.g. be skipped.)

Reply:
> the returned [com.google.common.base.Predicate](https://guava.dev/releases/32.1.1-jre/api/docs/com/google/common/base/Predicate.html) is described as a legacy type that should be avoided whenever possible.

good point, should that be 'attacked' even more generically then? essentially the entire `Predicates` class shouldn't be used then 😅

Reply:
Ideally yes! `git grep com.google.common.base.Predicate -- '*.java'` shows that internally not too many usages remain, so it's doable. But in some cases the result may be passed to a `com.google.common.base.Predicate`-expecting method, in which case a Refaster approach isn't great. That's the same issue as we have here, but I expect that the impact will be larger when replacing `or` and `not`. So propose to defer that to a separate PR.